File: chatcli/fetchers/grok.py
"""Grok 对话 fetcher：通过 CDP 抓取 grok.com 当前对话并投影为 ChatCli 消息格式。

对外暴露两个函数：
  - fetch_grok_conversation(port=None, url=None) -> dict   纯操作，抛异常
  - fetch_and_normalize(port=None, url=None) -> list | None  分发器友好版

参数：
  port: 本地 Chrome 的远程调试端口，默认 9222（沿用 config.cdp_port）
  url:  可选；提供时会让 Chrome 打开新标签并跳转到该地址，再在那个标签里
        抽取对话。grok.com 需要登录态，必须复用本机 Chrome 的用户配置。
"""

from importlib import resources

from chatcli import cdp
from chatcli.cdp import CDPError


_JS_PACKAGE = "chatcli.fetchers"
_JS_NAME = "grok.js"


def _load_js_source() -> str:
    """从包内资源读取 grok.js 源码。

    使用 stdlib importlib.resources（Python 3.9+），在 editable install
    和 wheel install 下都能工作，前提是 pyproject.toml 里把 fetchers/*.js
    写进 package-data。
    """
    return (
        resources.files(_JS_PACKAGE)
        .joinpath(_JS_NAME)
        .read_text(encoding="utf-8")
    )


def fetch_grok_conversation(port=None, url=None) -> dict:
    """连接本地 Chrome，注入 grok.js，返回 Grok 原始导出格式的 dict。

    url:  None 时使用当前已打开的 grok 对话标签；
          字符串时让 Chrome 新开一个标签并跳过去（grok.js 会自等待加载）。

    Raises:
        chatcli.cdp.CDPError 及子类（CDPRefused / CDPNoTab / CDPEvalError）
    """
    if url:
        ws_url = cdp.open_new_tab_ws(port=port, url=url)
    else:
        ws_url = cdp.find_grok_tab_ws(port=port)
    expression = _load_js_source()
    # awaitPromise=True 让 grok.js 的 polling/wait 真正生效（CDP 会等到 Promise resolve）
    # 开新标签 + 等渲染 + 等消息一起算，单次允许更长时间
    return cdp.evaluate_and_parse(
        ws_url,
        expression,
        timeout=30.0,
        await_promise=True,
    )


def fetch_and_normalize(port=None, url=None):
    """fetch_grok_conversation 的分发器友好版。

    返回 ChatCli 期望的 [{role, content}, ...]（role 为 user/assistant）。
    失败时打印中文错误并返回 None。
    """
    try:
        payload = fetch_grok_conversation(port=port, url=url)
    except CDPError as e:
        print(f"错误: {e}")
        return None
    except Exception as e:
        print(f"错误: 从 grok.com 抓取失败: {e}")
        return None

    result = []
    for m in payload.get("messages") or []:
        role = m.get("role")
        content = (m.get("content") or "").strip()
        if role == "human":
            role = "user"
        elif role == "ai":
            role = "assistant"
        else:
            # 未知 role 跳过，避免污染上下文
            continue
        if not content:
            continue
        result.append({"role": role, "content": content})
    return result

File: chatcli/fetchers/grok.js
/**
 * Grok 对话页提取脚本（CDP 注入版）。
 *
 * 由 Python 端通过 Chrome DevTools Protocol 的 Runtime.evaluate 注入到
 * grok.com 对话页运行。提取 human / ai 消息后，把整个 payload 用
 * JSON.stringify 包成字符串返回，保证通过 CDP 序列化时不丢字段。
 *
 * 选择器与逻辑与原 Grok-Exporter 扩展的 content.js 保持一致，便于对照维护。
 *
 * 当 Python 端开了新标签再跳转到 URL 时，页面刚开始加载、消息还没渲染；
 * 配合 Runtime.evaluate 的 awaitPromise=True，本脚本可以异步轮询，等到
 * 至少抽出一条消息再返回。
 */

(function () {
  "use strict";

  const USER_SEL = '[data-testid="user-message"]';
  const AI_SEL = '[data-testid="assistant-message"]';
  const BOTH_SEL = USER_SEL + ", " + AI_SEL;
  const MD_SEL = ".response-content-markdown";
  const THINK_RE =
    /^(思考了\s*\d+\s*s|Thought for\s+\d+\s*s|Thinking(?:\s+for\s+\d+\s*s)?)\s*/i;

  // /json/new 后等待渲染的最长时间（毫秒）
  const MAX_WAIT_MS = 12000;
  const POLL_INTERVAL_MS = 300;

  function conversationIdFromUrl(href) {
    const m = String(href || "").match(/\/c\/([0-9a-fA-F-]{36})/);
    return m ? m[1] : null;
  }

  function cleanTitle(title) {
    return String(title || "")
      .replace(/\s*[-|]\s*Grok\s*$/i, "")
      .trim();
  }

  function extractThinkingLabel(el) {
    const btn = Array.from(el.querySelectorAll("button")).find((b) =>
      /思考了|Thought for|Thinking/i.test(b.innerText || "")
    );
    if (btn) return (btn.innerText || "").trim();
    const m = (el.innerText || "").match(
      /^(思考了\s*\d+\s*s|Thought for\s+\d+\s*s)/i
    );
    return m ? m[1] : null;
  }

  function extractMessages() {
    const nodes = Array.from(document.querySelectorAll(BOTH_SEL));
    return nodes.map((el, index) => {
      const testid = el.getAttribute("data-testid");
      const role = testid === "user-message" ? "human" : "ai";
      const md = el.querySelector(MD_SEL);
      let content = (md ? md.innerText : el.innerText || "").trim();
      let thinking_label = null;

      if (role === "ai") {
        thinking_label = extractThinkingLabel(el);
        if (thinking_label && content.startsWith(thinking_label)) {
          content = content.slice(thinking_label.length).replace(/^\n+/, "").trim();
        } else {
          content = content.replace(THINK_RE, "").trim();
        }
      }

      const msg = { index: index, role: role, content: content };
      if (role === "ai" && thinking_label) {
        msg.thinking_label = thinking_label;
      }
      return msg;
    });
  }

  function buildExportPayload() {
    const messages = extractMessages();
    const human_n = messages.filter((m) => m.role === "human").length;
    const ai_n = messages.filter((m) => m.role === "ai").length;
    return {
      conversation_id: conversationIdFromUrl(location.href),
      url: location.href.split("?")[0],
      title: cleanTitle(document.title),
      exported_at: new Date().toISOString(),
      message_count: messages.length,
      human_message_count: human_n,
      ai_message_count: ai_n,
      messages: messages,
    };
  }

  function ensureOnConversation() {
    // 在等待循环里也要校验，否则登录页 / about:blank 不会抛错
    if (!/\/c\/[0-9a-fA-F-]{36}/.test(location.pathname)) {
      return {
        ok: false,
        error: "当前不是对话页。请打开 https://grok.com/c/<id> 后再抓取。",
      };
    }
    return { ok: true };
  }

  function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  // 入口：先确保 URL 正确，再轮询直到消息出现或超时。
  // 用 throw 而不是返回 {ok:false}，因为 CDP Runtime.evaluate 会把 throw
  // 自动包装成 exceptionDetails，Python 端 CDPEvalError 直接拿到错误信息。
  return (async function () {
    const deadline = Date.now() + MAX_WAIT_MS;
    let lastPayload = null;

    while (true) {
      const ok = ensureOnConversation();
      if (!ok.ok) {
        // 还没跳到对话页就继续等
        if (Date.now() >= deadline) {
          throw new Error(ok.error);
        }
        await sleep(POLL_INTERVAL_MS);
        continue;
      }
      const payload = buildExportPayload();
      if (payload.messages.length) {
        return JSON.stringify(payload);
      }
      lastPayload = payload;
      if (Date.now() >= deadline) {
        throw new Error(
          "未找到消息。请确认对话已加载完成，或向下/向上滚动后再试。"
        );
      }
      await sleep(POLL_INTERVAL_MS);
    }
  })();
})();